import "ydb/core/protos/bind_channel_storage_pool.proto";
import "ydb/public/api/protos/ydb_cms.proto";
import "ydb/library/login/protos/login.proto";

package NKikimrSubDomains;

message TDomainKey {
    optional fixed64 SchemeShard = 1;
    optional fixed64 PathId = 2;
}

message TSubDomainSettings {
    optional uint64 PlanResolution = 1;
    optional uint64 Coordinators = 2;
    optional uint64 Mediators = 3;
    optional string Name = 4;
    optional uint32 TimeCastBucketsPerMediator = 5;
    repeated NKikimrStoragePool.TStoragePool StoragePools = 6;
    optional bool ExternalSchemeShard = 7 [default = false];
    optional bool ExternalHive = 8 [default = false];
    optional TDomainKey ResourcesDomainKey = 9;
    optional EServerlessComputeResourcesMode ServerlessComputeResourcesMode = 15;
    optional bool ExternalSysViewProcessor = 10 [default = false];
    optional TSchemeQuotas DeclaredSchemeQuotas = 11;
    optional Ydb.Cms.DatabaseQuotas DatabaseQuotas = 12;
    optional TAuditSettings AuditSettings = 13;
    optional bool ExternalStatisticsAggregator = 14 [default = false];
}

message TProcessingParams {
    optional uint32 Version = 1;  //label for track changes
    optional uint64 PlanResolution = 2;
    repeated fixed64 Coordinators = 3;
    optional uint64 TimeCastBucketsPerMediator = 4;
    repeated fixed64 Mediators = 5;

    optional fixed64 SchemeShard = 6;
    optional fixed64 Hive = 7;
    optional fixed64 SysViewProcessor = 8;
    optional fixed64 StatisticsAggregator = 10;

    //put there SubSchemeShard and SubHive at the future

    // Plan resolution for idle coordinators
    optional uint64 IdlePlanResolution = 9;
}

message TDiskSpaceUsage {
    message TTables {
        // in bytes
        optional uint64 TotalSize = 1;
        optional uint64 DataSize = 2;
        optional uint64 IndexSize = 3;
    }

    message TTopics {
        // in bytes
        optional uint64 ReserveSize = 1;
        optional uint64 AccountSize = 2;
        optional uint64 DataSize = 3;
        optional uint64 UsedReserveSize = 4;
    }

    optional TTables Tables = 1;
    optional TTopics Topics = 2;
}

message TDomainState {
    optional bool DiskQuotaExceeded = 1;
}

message TAuditSettings {
    optional bool EnableDmlAudit = 1 [default = false];
    repeated string ExpectedSubjects = 2;
}

message TDomainDescription {
    optional fixed64 SchemeShardId_Depricated = 1;
    optional fixed64 PathId_Depricated = 2;

    optional TProcessingParams ProcessingParams = 3;
    optional TDomainKey DomainKey = 4; // SchemeShardId:PathId -- key for subdomain
    repeated NKikimrStoragePool.TStoragePool StoragePools = 5;

    optional uint64 PathsInside = 6;
    optional uint64 PathsLimit = 7;
    optional uint64 ShardsInside = 8;
    optional uint64 ShardsLimit = 9;

    optional TDomainKey ResourcesDomainKey = 10;
    optional EServerlessComputeResourcesMode ServerlessComputeResourcesMode = 23;

    optional TDiskSpaceUsage DiskSpaceUsage = 11; // TODO: temp for SLYDB-95

    optional TSchemeQuotas DeclaredSchemeQuotas = 12;

    optional uint64 PQPartitionsInside = 13;
    optional uint64 PQPartitionsLimit = 14;

    optional Ydb.Cms.DatabaseQuotas DatabaseQuotas = 15;
    optional TDomainState DomainState = 16;

    optional NLoginProto.TSecurityState SecurityState = 20;

    optional TAuditSettings AuditSettings = 21;
    optional fixed64 SharedHive = 22;
}

message TSchemeQuotas {
    repeated TSchemeQuota SchemeQuotas = 1;
}

message TSchemeQuota {
    optional double BucketSize = 1;
    optional uint64 BucketSeconds = 2;
}

enum EServerlessComputeResourcesMode {
    SERVERLESS_COMPUTE_RESOURCES_MODE_UNSPECIFIED = 0;
    SERVERLESS_COMPUTE_RESOURCES_MODE_SHARED = 1;     // using nodes that serve resources domain
    SERVERLESS_COMPUTE_RESOURCES_MODE_DEDICATED = 2;  // using nodes that serve self domain
}
